src/gt4py/next/iterator/transforms/global_tmps.py

```diff
@@ -19,11 +20,94 @@
     domain_utils,
     ir_makers as im,
 )
+from gt4py.next.iterator.ir_utils.domain_utils import SymbolicDomain
 from gt4py.next.iterator.transforms import cse, infer_domain, inline_lambdas
 from gt4py.next.iterator.type_system import inference as type_inference
 from gt4py.next.type_system import type_info, type_specifications as ts
 
 
+def select_elems_by_domain(
+    select_domain: SymbolicDomain,
+    target: itir.Expr,
+    args: Sequence[itir.Expr],
+    domains: tuple[SymbolicDomain, ...],
+):
+    """
+    Select all elements of possibly nested tuples for the given domain.
+
+    Returns (non-nested) tuples of the selected elements and the corresponding targets.
+    """
+    new_targets = []
+    new_els = []
+    for i, (el, el_domain) in enumerate(zip(args, domains)):
+        current_target = im.tuple_get(i, target)
+        if isinstance(el_domain, tuple):
+            assert cpm.is_call_to(el, "make_tuple")
+            more_targets, more_els = select_elems_by_domain(
+                select_domain, current_target, el.args, el_domain
+            )
+            new_els.extend(more_els)
+            new_targets.extend(more_targets)
+        else:
+            assert isinstance(el_domain, SymbolicDomain)
+            if el_domain == select_domain:
+                new_targets.append(current_target)
+                new_els.append(el)
+    return new_targets, new_els
+
+
+def _set_at_for_domain(stmt: itir.SetAt, domain: SymbolicDomain) -> itir.SetAt:
+    """Extract all elements with given domain into a new `SetAt` statement."""
+    tuple_expr = stmt.expr
+    assert cpm.is_call_to(tuple_expr, "make_tuple")
+    targets, expr_els = select_elems_by_domain(
+        domain, stmt.target, tuple_expr.args, stmt.expr.annex.domain
+    )
+    new_expr = im.make_tuple(*expr_els)
+    new_expr.annex.domain = domain
+
+    return itir.SetAt(expr=new_expr, domain=domain.as_expr(), target=im.make_tuple(*targets))
+
+
+def _populate_and_homogenize_domains(stmts: list[itir.Stmt]) -> list[itir.Stmt]:
+    """
+    Splits `SetAt` statements with multiple domains into multiple statements.
+
+    `SetAt`s have a single domain therefore the target domains have to be the same.
+    We support nested tuples by flattening them and recombining them into non-nested tuples.
+    """
+    new_stmts: list[itir.Stmt] = []
+    for stmt in stmts:
+        if isinstance(stmt, itir.SetAt):
+            assert hasattr(stmt.expr.annex, "domain")
+            distinct_domains: list[domain_utils.SymbolicDomain] = list(
+                # ordered set for reproducibility
+                dict.fromkeys(next_utils.flatten_nested_tuple(stmt.expr.annex.domain))
```

Comment:
I'm a bit confused here: why is guaranteed that the list is always sorted in the same way? Doesn't the `dict` created with `fromkeys()` depend on the the iteration order of whatever is in the flattened version of the `annex.domain`? Are we sure then that the contents of `annex.domain` are always sorted consistently?

Reply:
It's only important the for the same program the order is always the same (for test reproducibility).